src/folding/protogalaxy/folding.rs

```diff
@@ -0,0 +1,622 @@
+use ark_crypto_primitives::sponge::Absorb;
+use ark_ec::{CurveGroup, Group};
+use ark_ff::PrimeField;
+use ark_poly::{
+    univariate::{DensePolynomial, SparsePolynomial},
+    DenseUVPolynomial, EvaluationDomain, Evaluations, GeneralEvaluationDomain, Polynomial,
+};
+use ark_std::log2;
+use ark_std::{cfg_into_iter, Zero};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use std::marker::PhantomData;
+use std::ops::Add;
+
+use super::traits::ProtoGalaxyTranscript;
+use super::utils::{all_powers, bit_decompose, powers_of_beta};
+use super::ProtoGalaxyError;
+use super::{CommittedInstance, Witness};
+
+use crate::ccs::r1cs::R1CS;
+use crate::transcript::Transcript;
+use crate::utils::vec::*;
+use crate::Error;
+
+#[derive(Clone, Debug)]
+/// Implements the protocol described in section 4 of
+/// [ProtoGalaxy](https://eprint.iacr.org/2023/1106.pdf)
+pub struct Folding<C: CurveGroup> {
+    _phantom: PhantomData<C>,
+}
+impl<C: CurveGroup> Folding<C>
+where
+    <C as Group>::ScalarField: Absorb,
+    <C as CurveGroup>::BaseField: Absorb,
+{
+    #![allow(clippy::type_complexity)]
+    /// implements the non-interactive Prover from the folding scheme described in section 4
+    pub fn prove(
+        transcript: &mut (impl Transcript<C> + ProtoGalaxyTranscript<C>),
+        r1cs: &R1CS<C::ScalarField>,
+        // running instance
+        instance: &CommittedInstance<C>,
+        w: &Witness<C>,
+        // incomming instances
+        vec_instances: &[CommittedInstance<C>],
+        vec_w: &[Witness<C>],
+    ) -> Result<
+        (
+            CommittedInstance<C>,
+            Witness<C>,
+            Vec<C::ScalarField>,
+            Vec<C::ScalarField>,
+        ),
+        Error,
+    > {
+        let t = instance.betas.len();
+        let n = r1cs.A.n_cols;
+        if w.w.len() != n {
+            return Err(Error::NotSameLength(
+                "w.w.len()".to_string(),
+                w.w.len(),
+                "n".to_string(),
+                n,
+            ));
+        }
+        if log2(n) as usize != t {
+            return Err(Error::NotEqual);
+        }
+
+        // absorb the committed instances
+        transcript.absorb_committed_instance(instance)?;
+        for ci in vec_instances.iter() {
+            transcript.absorb_committed_instance(ci)?;
+        }
+
+        let delta = transcript.get_challenge();
+        let deltas = powers_of_beta(delta, t);
+
+        let f_w = eval_f(r1cs, &w.w)?;
+
+        // F(X)
+        let mut F_X: SparsePolynomial<C::ScalarField> = SparsePolynomial::zero();
+        for (i, f_w_i) in f_w.iter().enumerate() {
+            let lhs = pow_i_over_x::<C::ScalarField>(i, &instance.betas, &deltas)?;
+            let curr = &lhs * *f_w_i;
+            F_X = F_X.add(curr);
+        }
+
+        let F_X_dense = DensePolynomial::from(F_X.clone());
+        transcript.absorb_vec(&F_X_dense.coeffs);
```

Comment:
The F_X_dense is later returned out at the end of the function, it should have $t$ coeffs, where $t=log(n)$, where $n$ is the witness length, so not a big number of coefficients. And I think that probably they are not zero, since it is computed from $\sum_{i \in [n]} pow_i(\overrightarrow{\beta} + X \cdot \overrightarrow{\delta}) \cdot f_i(w)$, where the $pow_i(\overrightarrow{\beta} + X \cdot \overrightarrow{\delta})$ part results into polynomials of degree of maximum $log(i)$ that get combined with $f_i(w)$, which should be zero only in the 'native' case (the non folded case).

Reply:
Ohhh that's true!!! Could you add some comments about that? I think it's important to reflect it JIC we later look for optimizations :)